### PR TITLE
Update: add comment to message in no-warning-comments (fixes #12327)

### LINKED ISSUE #12327: `no-warning-comments` should include the comment itself in the report (UX only)
I think the point of this rule is to keep track of TODO comments, however this is hard to follow; I'd have to open each file separately to see if I can handle a task.

```ts
  source/globals.d.ts:1:1
  ⚠    1:1   Unexpected todo comment.                                                                   no-warning-comments
  ⚠    2:1   Unexpected todo comment.                                                                   no-warning-comments

  source/libs/post-form.ts:9:3
  ⚠    9:3   Unexpected todo comment.                                                                   no-warning-comments

  source/features/hidden-review-comments-indicator.tsx:37:2
  ⚠   37:2   Unexpected todo comment.                                                                   no-warning-comments

  source/features/revert-file.tsx:89:4
  ⚠   89:4   Unexpected todo comment.                                                                   no-warning-comments
```

**What rule do you want to change?**

`no-warning-comments`

**Does this change cause the rule to produce more or fewer warnings?**

Unchanged. It's only about UX

**How will the change be implemented? (New option, new default behavior, etc.)?**

New default

**What does the rule currently do for this code?**

Shown above

**What will the rule do after it's changed?**

```ts
  source/globals.d.ts:1:1
  ⚠    1:1   FIXME: Drop ignore after #9.                                                                  no-warning-comments
  ⚠    2:1   FIXME: Find this info on the page.                                                                   no-warning-comments

  source/libs/post-form.ts:9:3
  ⚠    9:3   TODO: Improve JSX types for event listeners so we can use... [clipped]                                                                   no-warning-comments

  source/features/hidden-review-comments-indicator.tsx:37:2
  ⚠   37:2   TODO: Drop some definitions when their related bugs are resolved                                                                   no-warning-comments

  source/features/revert-file.tsx:89:4
  ⚠   89:4   TODO: Drop `as` after https://github.com/microsoft/TSJS-lib-generator/issues/741                                                                   no-warning-comments
```

### PR DESCRIPTION
**Pull request by anikethsaha**
#### Prerequisites checklist

- [X] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/master/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[X] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)
Added data in the error message which will show the comment itself. 
It will only show till 30 characters

image

#### Is there anything you'd like reviewers to focus on?
Is `30` is fine ? after `30` char it will be truncated.

### CODE REVIEW

**Comment by mdjermanovic** [comment]
Should we avoid linebreaks in multiline comments, for example for [this one in linter.js](https://github.com/eslint/eslint/blob/e71e2980cd2e319afc70d8c859c7ffd59cf4157b/lib/linter/linter.js#L542) the rule would output:

```bash
542:5  error  Unexpected 'todo' comment: * TODO: @aladdin-add
     * 1...  no-warning-comments
```

Also, if it isn't too difficult, it might be nice to avoid cutting words, surrogate pairs, and composite characters?

File: lib/rules/no-warning-comments.js [hunk 1]
```diff
@@ -141,12 +141,18 @@ module.exports = {
 
             const matches = commentContainsWarningTerm(node.value);
 
-            matches.forEach(matchedTerm => {
+            matches.forEach(({ type, comment }) => {
+                let commentToDisplay = comment.trim();
+
+                if (commentToDisplay.length >= 30) {
```
Comment (mdjermanovic) [inline lib/rules/no-warning-comments.js:147]: It would be nice to declare a constant for `30`.

File: lib/rules/no-warning-comments.js [hunk 2]
```diff
@@ -141,12 +141,18 @@ module.exports = {
...
+                    commentToDisplay = `${commentToDisplay.slice(0, 30)}..`;
```
Comment (mdjermanovic) [inline lib/rules/no-warning-comments.js:148]: Maybe `...` instead of `..`?

Though, I guess this is language-specific punctuation, so it might be technically correct to create a new `messageId` for this case.

File: lib/rules/no-warning-comments.js (hunk 1 above)
Comment (anikethsaha) [reply lib/rules/no-warning-comments.js:147]: cool

File: lib/rules/no-warning-comments.js (hunk 2 above)
Comment (anikethsaha) [reply lib/rules/no-warning-comments.js:148]: there is an `.` at the end of the message. so if I am adding `...` it would be `....` then. 
Thats why.

> so it might be technically correct to create a new `messageId` for this case.

actually I am using the existing one as I dont think a new one is needed. if we are putting this update an option, I guess we can then make separate message ids.

**Comment by anikethsaha** [comment]
> Should we avoid linebreaks in multiline comments, for example for [this one in linter.js](https://github.com/eslint/eslint/blob/e71e2980cd2e319afc70d8c859c7ffd59cf4157b/lib/linter/linter.js#L542) the rule would output:
> 
> ```shell
> 542:5  error  Unexpected 'todo' comment: * TODO: @aladdin-add
>      * 1...  no-warning-comments
> ```

Or can we add `\n` when there is a linebreak ?

> Also, if it isn't too difficult, it might be nice to avoid cutting words, surrogate pairs, and composite characters?

I didn't get this

**Comment by mdjermanovic** [comment]
> > Should we avoid linebreaks in multiline comments, for example for [this one in linter.js](https://github.com/eslint/eslint/blob/e71e2980cd2e319afc70d8c859c7ffd59cf4157b/lib/linter/linter.js#L542) the rule would output:
> > ```shell
> > 542:5  error  Unexpected 'todo' comment: * TODO: @aladdin-add
> >      * 1...  no-warning-comments
> > ```
> 
> Or can we add `\n` when there is a linebreak ?

Sorry, I didn't understand this, how would the output look like? I think we're usually trying to avoid linebreaks in final messages. In this case, the final message has a linebreak taken from the source code.

>> Also, if it isn't too difficult, it might be nice to avoid cutting words, surrogate pairs, and composite characters?

> I didn't get this

For example, for [this comment](https://github.com/eslint/eslint/blob/e71e2980cd2e319afc70d8c859c7ffd59cf4157b/lib/rules/no-warning-comments.js#L70), should the message be:

`Unexpected 'todo' comment: * If the term ends in a word c...`

or

`Unexpected 'todo' comment: * If the term ends in a word...`

Also, `slice` might accidentally cut some code units that should be together:

```js
"👍".slice(0, 1) // �
```

File: lib/rules/no-warning-comments.js (hunk 2 above)
Comment (mdjermanovic) [reply lib/rules/no-warning-comments.js:148]: Ah, it seems that the default formatter doesn't show the last `.` in messages. We should maybe put quotes around the comment (and `...` in quotes) to avoid this.

> > so it might be technically correct to create a new `messageId` for this case.
> 
> actually I am using the existing one as I dont think a new one is needed. if we are putting this update an option, I guess we can then make separate message ids.

It might be nice for translation purposes if `...` isn't common for all languages (I'm not sure about this), though I guess we could improve that later if needed.

Comment (anikethsaha) [reply lib/rules/no-warning-comments.js:148]: > Ah, it seems that the default formatter doesn't show the last `.` in messages. We should maybe put quotes around the comment (and `...` in quotes) to avoid this.

Like this 
```
542:5  error  Unexpected 'todo' comment: "* TODO: @aladdin-add...".
```

Comment (anikethsaha) [reply lib/rules/no-warning-comments.js:148]: > It might be nice for translation purposes if `...` isn't common for all languages (I'm not sure about this), though I guess we could improve that later if needed.

like js, ts it might be common (not sure though).
any other lang we need to consider ? may be they are out of scope for eslint

**Comment by anikethsaha** [comment]
> Sorry, I didn't understand this, how would the output look like? I think we're usually trying to avoid linebreaks in final messages. In this case, the final message has a linebreak taken from the source code.

like this I meant 

```js
comment.replace('\n', '\\n')
```

So it will be like 

```
542:5  error  Unexpected 'todo' comment: * TODO: @aladdin-add \n  * 1...  no-warning-comments
```

WDYT ?

> For example, for [this comment](https://github.com/eslint/eslint/blob/e71e2980cd2e319afc70d8c859c7ffd59cf4157b/lib/rules/no-warning-comments.js#L70), should the message be:
> 
> `Unexpected 'todo' comment: * If the term ends in a word c...`
> 
> or
> 
> `Unexpected 'todo' comment: * If the term ends in a word...`
> 
> Also, `slice` might accidentally cut some code units that should be together:
> 
> ```js
> "👍".slice(0, 1) // �
> ```

OKay, I will try to implement the same without slice

File: lib/rules/no-warning-comments.js (hunk 2 above)
Comment (mdjermanovic) [reply lib/rules/no-warning-comments.js:148]: > > It might be nice for translation purposes if `...` isn't common for all languages (I'm not sure about this), though I guess we could improve that later if needed.
> 
> like js, ts it might be common (not sure though).
> any other lang we need to consider ? may be they are out of scope for eslint

I meant languages other than English, those that have some other punctuation that indicates an intentional omission, but we can audit all the rules and fix things like this later when we precisely define the infrastructure for localization of error messages.

Comment (mdjermanovic) [reply lib/rules/no-warning-comments.js:148]: > Like this ?
> 
> ```
> 542:5  error  Unexpected 'todo' comment: "* TODO: @aladdin-add...".
> ```

Yes, like that, just with single quotes (I think we're aways using single quotes in error messages).

**Comment by mdjermanovic** [comment]
Maybe we could split the comment by whitespace, and then concatenate the parts with 1 space as a separator while the total length doesn't exceed the max?

I guess that would solve all problems with linebreaks, breaking words, surrogate pairs, etc. The comment wouldn't look exactly like in the source code, but the significant content would be shown.

File: lib/rules/no-warning-comments.js (hunk 2 above)
Comment (anikethsaha) [reply lib/rules/no-warning-comments.js:148]: > I meant languages other than English,

ops,  sorry 😄 

> those that have some other punctuation that indicates an intentional omission, but we can audit all the rules and fix things like this later when we precisely define the infrastructure for localization of error messages.

yeah, cool

> Yes, like that, just with single quotes (I think we're aways using single quotes in error messages).

cool

**Comment by anikethsaha** [comment]
> Maybe we could split the comment by whitespace, and then concatenate the parts with 1 space as a separator while the total length doesn't exceed the max?
> 
> I guess that would solve all problems with linebreaks, breaking words, surrogate pairs, etc. The comment wouldn't look exactly like in the source code, but the significant content would be shown.

yes, I have the similar thought. 
Though I was thinking of adding `\n` as mentioned above comment in the linebreak so that the user can know that there is a linebreak instead of simply copying the text and searching for a particular comment in the file.

WDYT ?

**Comment by mdjermanovic** [comment]
> Though I was thinking of adding `\n` as mentioned above comment in the linebreak so that the user can know that there is a linebreak instead of simply copying the text and searching for a particular comment in the file.

I don't have a strong opinion on this. It might be useful for someone (though we also report comment's line), but it might be seen as extra characters that negatively affect readability by someone else.

File: lib/rules/no-warning-comments.js (hunk 2 above)
Comment (anikethsaha) [reply lib/rules/no-warning-comments.js]: done

File: lib/rules/no-warning-comments.js [hunk 3]
```diff
@@ -8,6 +8,8 @@
 const { escapeRegExp } = require("lodash");
 const astUtils = require("./utils/ast-utils");
 
+const CHAR_LIMIT = 30;
```
Comment (mdjermanovic) [inline lib/rules/no-warning-comments.js:11]: I think this can be at least `40`, maybe even more. Thoughts?

File: lib/rules/no-warning-comments.js [hunk 4]
```diff
@@ -141,12 +143,28 @@ module.exports = {
...
             const matches = commentContainsWarningTerm(node.value);
 
-            matches.forEach(matchedTerm => {
+            matches.forEach(({ type, comment }) => {
+                let len = 0;
+                const comments = [];
+                let truncated = false;
+
+                comment
+                    .split(/\s/u).forEach(c => {
```
Comment (mdjermanovic) [inline lib/rules/no-warning-comments.js:152]: Maybe we could do `/\s+/u` instead, to turn a sequence of whitespace into just one space?

Multiline comments will often have some indentation after a linebreak, so it might be nice to not show all that blank space in the message?

File: lib/rules/no-warning-comments.js [hunk 5]
```diff
@@ -141,12 +143,28 @@ module.exports = {
...
+                        if (c.length + len <= CHAR_LIMIT) {
+                            comments.push(c);
+                        } else {
+                            truncated = true;
```
Comment (mdjermanovic) [inline lib/rules/no-warning-comments.js:156]: It might be good to break the loop here

File: lib/rules/no-warning-comments.js [hunk 6]
```diff
@@ -141,12 +143,28 @@ module.exports = {
...
+                        }
+
+                        len += c.length;
```
Comment (mdjermanovic) [inline lib/rules/no-warning-comments.js:159]: Should we also count spaces that will be added between?

File: lib/rules/no-warning-comments.js [hunk 7]
```diff
@@ -42,7 +44,7 @@ module.exports = {
         ],
 
         messages: {
-            unexpectedComment: "Unexpected '{{matchedTerm}}' comment."
+            unexpectedComment: "Unexpected '{{type}}' comment: {{comment}}."
```
Comment (mdjermanovic) [inline lib/rules/no-warning-comments.js:47]: I think we usually add `''` quotes in the message. Like `'{{type}}'`, it could be `'{{comment}}'`?

File: lib/rules/no-warning-comments.js [hunk 8]
```diff
@@ -141,12 +143,28 @@ module.exports = {
 
             const matches = commentContainsWarningTerm(node.value);
 
-            matches.forEach(matchedTerm => {
+            matches.forEach(({ type, comment }) => {
+                let len = 0;
+                const comments = [];
```
Comment (mdjermanovic) [inline lib/rules/no-warning-comments.js:148]: Could we rename this array? `comments` is a bit confusing, since the array contains parts of one comment.

File: lib/rules/no-warning-comments.js (hunk 3 above)
Comment (anikethsaha) [reply lib/rules/no-warning-comments.js:11]: I think `40` is fine.

File: lib/rules/no-warning-comments.js (hunk 4 above)
Comment (anikethsaha) [reply lib/rules/no-warning-comments.js:152]: cool

File: lib/rules/no-warning-comments.js (hunk 6 above)
Comment (anikethsaha) [reply lib/rules/no-warning-comments.js:159]: Actually, I am counting only the readable char. 

We can count the number of space, (`comments.lenght - 1`)

File: lib/rules/no-warning-comments.js [hunk 9]
```diff
@@ -141,12 +143,29 @@ module.exports = {
 
             const matches = commentContainsWarningTerm(node.value);
 
-            matches.forEach(matchedTerm => {
+            matches.forEach(({ type, comment }) => {
+                let len = 0;
+                const commentsToDisplay = [];
```
Comment (mdjermanovic) [inline lib/rules/no-warning-comments.js:148]: Thoughts about building the final string instead of an array? That way we could avoid one extra variable `len`.

File: lib/rules/no-warning-comments.js [hunk 10]
```diff
@@ -141,12 +143,25 @@ module.exports = {
 
             const matches = commentContainsWarningTerm(node.value);
 
-            matches.forEach(matchedTerm => {
+            matches.forEach(({ type, comment }) => {
+                let commentsToDisplay = "";
```
Comment (mdjermanovic) [inline lib/rules/no-warning-comments.js:147]: Can we name this `commentToDisplay`

File: lib/rules/no-warning-comments.js [hunk 11]
```diff
@@ -141,12 +143,25 @@ module.exports = {
...
+                let truncated = false;
+
+                for (const c of comment.trim().split(/\s+/u)) {
+                    commentsToDisplay = commentsToDisplay.trim();
+                    if (c.length + commentsToDisplay.length <= CHAR_LIMIT) {
```
Comment (mdjermanovic) [inline lib/rules/no-warning-comments.js:152]: This can be off-by-one.

Maybe we could first make the next string, then check its length and if it's okay assign it to `commentToDisplay`.

Comment (anikethsaha) [reply lib/rules/no-warning-comments.js:152]: like this using a temporary variable ?

```js
for (const c of comment.trim().split(/\s+/u)) {
  commentToDisplay = commentToDisplay.trim();
  const tmp = `${commentToDisplay} ${c}`;

  if (tmp.length <= CHAR_LIMIT) {
    commentToDisplay = tmp;
  } else {
    truncated = true;
  }
}
```

Comment (mdjermanovic) [reply lib/rules/no-warning-comments.js:152]: Yes, something like that

Comment (anikethsaha) [reply lib/rules/no-warning-comments.js:152]: if we construct the next string first as mentioned above in the snippet, there may be cases where the `...` is being appended with a nonfull word i.e

instead of `and...`, there is a chance of having like `a...` as well in this [line](https://github.com/eslint/eslint/pull/13522/files#diff-0e787f8a02ecbdc84bc736003f9736ceR230)

I can refactor as this i.e extracting the condition to a variable. 

```js
for (const c of comment.trim().split(/\s+/u)) {
  commentToDisplay = commentToDisplay.trim();
  const len = commentToDisplay.length + c.length;

  if (len <= CHAR_LIMIT) {
    commentToDisplay = `${commentToDisplay} ${c}`;
  } else {
    truncated = true;
  }
}
```
WDYT ?

Comment (mdjermanovic) [reply lib/rules/no-warning-comments.js:152]: > if we construct the next string first as mentioned above in the snippet, there may be cases where the `...` is being appended with a nonfull word i.e
> 
> instead of `and...`, there is a chance of having like `a...` as well in this [line](https://github.com/eslint/eslint/pull/13522/files#diff-0e787f8a02ecbdc84bc736003f9736ceR230)

Hm, I'm not sure how could that happen if we're always appending full words

Comment (anikethsaha) [reply lib/rules/no-warning-comments.js:152]: yeah , 

image

Comment (mdjermanovic) [reply lib/rules/no-warning-comments.js:152]: That's probably the full word "a" that appears later in the comment, not the first letter of "and".

We should break the loop after `truncated = true`.

Comment (anikethsaha) [reply lib/rules/no-warning-comments.js]: done

File: lib/rules/no-warning-comments.js [hunk 12]
```diff
@@ -135,18 +144,39 @@ module.exports = {
...
 
             const matches = commentContainsWarningTerm(node.value);
 
-            matches.forEach(matchedTerm => {
+            matches.forEach(({ type, comment }) => {
+                let commentToDisplay = "";
+                let truncated = false;
+
+                for (const c of comment.trim().split(/\s+/u)) {
+                    commentToDisplay = commentToDisplay.trim();
```
Comment (mdjermanovic) [inline lib/rules/no-warning-comments.js:161]: I think `commentToDisplay = commentToDisplay.trim();` isn't necessary

File: lib/rules/no-warning-comments.js [hunk 13]
```diff
@@ -135,18 +144,39 @@ module.exports = {
...
+                    const tmp = `${commentToDisplay} ${c}`;
```
Comment (mdjermanovic) [inline lib/rules/no-warning-comments.js:162]: This will have one extra space before the first word (at least in the first iteration), so it could make the check bellow to be off-by-one.

Can we maybe do something like this:

```js
const tmp = commentToDisplay ? `${commentToDisplay} ${c}` : c;
```

File: lib/rules/no-warning-comments.js [hunk 14]
```diff
@@ -135,18 +144,39 @@ module.exports = {
...
+                    }
+                }
+
                 context.report({
                     node,
                     messageId: "unexpectedComment",
                     data: {
-                        matchedTerm
+                        type,
+                        comment: `${commentToDisplay.trim()}${
```
Comment (mdjermanovic) [inline lib/rules/no-warning-comments.js:177]: If we change `const tmp = commentToDisplay ? ...` per the previous comment, I think that this `trim` will be also unnecessary.

File: lib/rules/no-warning-comments.js [hunk 15]
```diff
@@ -122,7 +131,7 @@ module.exports = {
 
             warningRegExps.forEach((regex, index) => {
                 if (regex.test(comment)) {
-                    matches.push(warningTerms[index]);
+                    matches.push({ type: warningTerms[index], comment });
```
Comment (mdjermanovic) [inline lib/rules/no-warning-comments.js:134]: Is it necessary to add `comment` here? We could do  `const comment = node.value` in `function checkComment`

File: lib/rules/no-warning-comments.js [hunk 16]
```diff
@@ -135,18 +144,39 @@ module.exports = {
...
+                selfConfigRegEx.test(node.value)
+            ) {
                 return;
             }
 
             const matches = commentContainsWarningTerm(node.value);
 
             matches.forEach(matchedTerm => {
+                const comment = node.value;
+                let commentToDisplay;
```
[the remark this answers is not in the document]
Comment (anikethsaha) [reply lib/rules/no-warning-comments.js:158]: 👏👏👏
This is really a great edge case. Thanks a lot for pointing  👍

File: tests/lib/rules/no-warning-comments.js [hunk 17]
```diff
@@ -244,6 +244,19 @@ ruleTester.run("no-warning-comments", rule, {
                     }
                 }
             ]
+        },
+        {
+            code: "// TODO: https://github.com/eslint/eslint/pull/13522#discussion_r470293411",
```
Comment (mdjermanovic) [inline tests/lib/rules/no-warning-comments.js:249]: :+1: nice self-explanatory test :)

Just, I think that in order to test that edge case, the long URL should be the first word, before `TODO:`. In this example, it will get `TODO:` as the first word.

Comment (anikethsaha) [reply tests/lib/rules/no-warning-comments.js:249]: ohh shoot, yeah my bad.

**Review by mdjermanovic — APPROVED** [review]
Looks great, thanks!

**Comment by kaicataldo** [comment]
Thanks for contributing!

**Comment by fregante** [comment]
Awesome!
